src/coprocessor/dag/batch_executor/limit.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::interface::*;
+use crate::coprocessor::Result;
+use tipb::expression::FieldType;
+
+pub struct BatchLimitExecutor<Src: BatchExecutor> {
+    src: Src,
+    expect_rows: usize,
+}
+
+impl<Src: BatchExecutor> BatchLimitExecutor<Src> {
+    pub fn new(src: Src, limit: usize) -> Result<Self> {
+        if limit == 0 {
+            return Err(box_err!("limit should not be zero"));
+        }
+        Ok(Self {
+            src,
+            expect_rows: limit,
+        })
+    }
+}
+
+impl<Src: BatchExecutor> BatchExecutor for BatchLimitExecutor<Src> {
+    #[inline]
+    fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+        // we needn't check whether expect_rows is 0 since src will check it.
+        let mut result = if expect_rows > self.expect_rows {
+            self.src.next_batch(self.expect_rows)
+        } else {
+            self.src.next_batch(expect_rows)
+        };
+        if result.is_drained.is_err() {
```

Comment:
We still need to filter the result when result is drained because drained only means that there is no *further* data, doesn't mean that there is no data in this result.